internal/querycoord/segment_allocator.go

```diff
@@ -0,0 +1,241 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+package querycoord
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/commonpb"
+	"github.com/milvus-io/milvus/internal/proto/querypb"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+)
+
+const MaxMemUsagePerNode = 0.9
+
+func defaultSegAllocatePolicy() SegmentAllocatePolicy {
+	return shuffleSegmentsToQueryNodeV2
+}
+
+// SegmentAllocatePolicy helper function definition to allocate Segment to queryNode
+type SegmentAllocatePolicy func(ctx context.Context, collectionID UniqueID, reqs []*querypb.LoadSegmentsRequest, cluster Cluster, wait bool, excludeNodeIDs []int64) error
+
+// shuffleSegmentsToQueryNode shuffle segments to online nodes
+// returned are noded id for each segment, which satisfies:
+//     len(returnedNodeIds) == len(segmentIDs) && segmentIDs[i] is assigned to returnedNodeIds[i]
+func shuffleSegmentsToQueryNode(ctx context.Context, collectionID UniqueID, reqs []*querypb.LoadSegmentsRequest, cluster Cluster, wait bool, excludeNodeIDs []int64) error {
+	if len(reqs) == 0 {
+		return nil
+	}
+
+	maxNumSegments := 0
```

Comment:
We can have two heaps.
1st heap is min heap, 2nd heap is maxheap, always compare two heap top and see if we can move one segment to make the cluster balancer.